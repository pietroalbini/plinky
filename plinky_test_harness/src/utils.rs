use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
