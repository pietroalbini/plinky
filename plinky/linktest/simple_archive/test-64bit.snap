---
source: plinky_test_harness/src/utils.rs
---
linking exited with exit status: 0

no stdout present

=== stderr ===
debug print: loaded object
 │
 │  class: Elf64, endian: Little, abi: SystemV, machine: X86_64
 │
 │  section .comment (no perms) in <plinky>
 │   │
 │   │  zero-terminated strings should be deduplicated
 │   │
 │   │  ╭──────────────────────────────────────────────┬─────────────────╮
 │   │  │ 4c 69 6e 6b 65 72 3a 20 70 6c 69 6e 6b 79 00 │ Linker: plinky. │
 │   │  ╰──────────────────────────────────────────────┴─────────────────╯
 │   ┴
 │
 │  section .data#0 (perms: rw) in hello.o inside archive archived.a
 │   │
 │   │  ╭────────────────────────────────────────┬───────────────╮
 │   │  │ 48 65 6c 6c 6f 20 77 6f 72 6c 64 21 0a │ Hello world!. │
 │   │  ╰────────────────────────────────────────┴───────────────╯
 │   ┴
 │
 │  section names section .shstrtab in <plinky>
 │   │
 │   │  section names
 │   ┴
 │
 │  string table section .strtab in <plinky>
 │   │
 │   │  symbol names for: symbol table
 │   ┴
 │
 │  symbols table section .symtab in <plinky>
 │   │
 │   │  view: symbol table
 │   │  strings: .strtab
 │   ┴
 │
 │  section .text#0 (perms: rx) in entry.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ b8 04 00 00 00 bb 01 00 00 00 b9 00 00 00 00 ba │ ................ │
 │   │  │ 00 00 00 00 cd 80 b0 01 bb 00 00 00 00 cd 80    │ ...............  │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   │
 │   │    Relocations:
 │   │  ╭────────────┬───────────┬────────┬────────╮
 │   │  │ Type       │ Symbol    │ Offset │ Addend │
 │   │  ├────────────┼───────────┼────────┼────────┤
 │   │  │ Absolute32 │ hello     │ 0xb    │ 0x0    │
 │   │  ├────────────┼───────────┼────────┼────────┤
 │   │  │ Absolute32 │ hello_len │ 0x10   │ 0x0    │
 │   │  ╰────────────┴───────────┴────────┴────────╯
 │   ┴
 │
 │    Symbols:
 │  ╭───────────┬──────┬───────────────────────────────────┬────────────┬───────────────╮
 │  │ Name      │ Type │ Source                            │ Visibility │ Value         │
 │  ├───────────┼──────┼───────────────────────────────────┼────────────┼───────────────┤
 │  │ <empty>   │ none │ <plinky>                          │ local      │ <null>        │
 │  ├───────────┼──────┼───────────────────────────────────┼────────────┼───────────────┤
 │  │ _start    │ none │ entry.o                           │ global     │ .text#0 + 0x0 │
 │  ├───────────┼──────┼───────────────────────────────────┼────────────┼───────────────┤
 │  │ hello     │ none │ hello.o inside archive archived.a │ global     │ .data#0 + 0x0 │
 │  ├───────────┼──────┼───────────────────────────────────┼────────────┼───────────────┤
 │  │ hello_len │ none │ hello.o inside archive archived.a │ global     │ 0xd           │
 │  ╰───────────┴──────┴───────────────────────────────────┴────────────┴───────────────╯
 │
 │  inputs
 │   │
 │   │    entry.o
 │   │  ╭─────────────────────┬─────╮
 │   │  │ X86 ISA used        │     │
 │   │  ├─────────────────────┼─────┤
 │   │  │ x86 features 2 used │ x86 │
 │   │  ╰─────────────────────┴─────╯
 │   │
 │   │    hello.o inside archive archived.a
 │   │  ╭─────────────────────┬─────╮
 │   │  │ X86 ISA used        │     │
 │   │  ├─────────────────────┼─────┤
 │   │  │ x86 features 2 used │ x86 │
 │   │  ╰─────────────────────┴─────╯
 │   ┴
 ┴

==============

running exited with exit status: 0

=== stdout ===
Hello world!

no stderr present
